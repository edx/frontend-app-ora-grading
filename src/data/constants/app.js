import { getConfig } from '@edx/frontend-platform';

export const routePath = `${getConfig().PUBLIC_PATH}:courseId`;
export const locationId = window.location.pathname.slice(1);
