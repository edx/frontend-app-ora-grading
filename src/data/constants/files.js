import { StrictDict } from 'utils';

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
});

export default FileTypes;
