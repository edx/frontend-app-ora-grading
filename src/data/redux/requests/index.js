export { actions, reducer } from './reducer';
