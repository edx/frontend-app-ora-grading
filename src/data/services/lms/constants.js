import { StrictDict } from 'utils';

export const lockStatuses = StrictDict({
  unlocked: 'unlocked',
  locked: 'locked',
  inProgress: 'in-progress',
});

export const gradeStatuses = StrictDict({
  ungraded: 'ungraded',
  graded: 'graded',
});

export const gradingStatuses = StrictDict({
  ungraded: gradeStatuses.ungraded,
  graded: gradeStatuses.graded,
  locked: lockStatuses.locked,
  inProgress: lockStatuses.inProgress,
});

export const gradingStatusDisplay = StrictDict({
  [gradingStatuses.ungraded]: 'Ungraded',
  [gradingStatuses.locked]: 'Currently being graded by someone else',
  [gradingStatuses.graded]: 'Grading Complete',
  [gradingStatuses.inProgress]: 'You are currently grading this response',
});

export const feedbackRequirement = StrictDict({
  disabled: 'disabled',
  required: 'required',
  optional: 'optional',
});
