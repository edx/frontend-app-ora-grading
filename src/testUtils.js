// eslint-disable-next-line import/prefer-default-export
export const formatMessage = (msg, values) => {
  let message = msg.defaultMessage;
  if (values === undefined) { return message; }
  Object.keys(values).forEach(key => {
    // eslint-disable-next-line
    message = message.replace(`{${key}}`, values[key]);
  });
  return message;
};
