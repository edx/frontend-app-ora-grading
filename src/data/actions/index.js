import { StrictDict } from 'utils';

import app from './app';
import grading from './grading';
import submissions from './submissions';

export default StrictDict({
  app,
  grading,
  submissions,
});
