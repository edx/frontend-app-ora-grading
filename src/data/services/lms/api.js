import { StrictDict } from 'utils';
import { gradeStatuses, lockStatuses } from './constants';
import fakeData from './fakeData';

// import urls from './urls';
// import { pageSize, paramKeys } from './constants';
// import messages from './messages';
// import * as utils from './utils';
// const { get, post, stringifyUrl } = utils;

/*********************************************************************************
 * GET Actions
 *********************************************************************************/

const mockSuccess = (returnValFn) => (...args) => (
  new Promise((resolve) => resolve(returnValFn(...args)))
);

const mockFailure = (returnValFn) => (...args) => (
  new Promise((resolve, reject) => reject(returnValFn(...args)))
);

/**
 * get('/api/initialize', { ora_location, course_id? })
 * @return {
 *   oraMetadata: { name, prompt, type ('individual' vs 'team'), rubricConfig, fileUploadResponseConfig },
 *   courseMetadata: { courseOrg, courseName, courseNumber, courseId },
 *   submissions: {
 *     [submissionUUID]: {
 *       id: <submissionUUID>, (not currently used)
 *       username
 *       submissionUUID
 *       dateSubmitted (timestamp)
 *       gradeStatus (['ungraded', 'graded', 'locked', 'locked_by_you'?])
 *       grade: { pointsEarned, pointsPossible }
 *     },
 *     ...
 *   },
 * }
 */
const initializeApp = mockSuccess(() => ({
  oraMetadata: fakeData.oraMetadata,
  courseMetadata: fakeData.courseMetadata,
  submissions: fakeData.submissions,
}));

/**
 * get('/api/submission', { submissionUUID })
 * @return {
 *   submision: {
 *     gradeData,
 *     gradeStatus,
 *     response: { files: [{}], text: <html> },
 *   },
 * }
 */
const fetchSubmission = mockSuccess((submissionUUID) => {
  console.log({ fetchSubmission: { submissionUUID } });
  return fakeData.mockSubmission(submissionUUID);
});

/**
 * fetches the current grade, gradeStatus, and rubricResponse data for the given submission
 * get('/api/submissionStatus', { submissionUUID })
 *   @return {obj} submissionStatus object
 *   {
 *     gradeData,
 *     gradeStatus,
 *     lockStatus,
 *   }
 */
const fetchSubmissionStatus = mockSuccess((submissionUUID) => (
  fakeData.mockSubmissionStatus(submissionUUID)
));

/**
 * Fetches only the learner response for a given submission. Used for pre-fetching response
 * for neighboring submissions in the queue.
 */
export const fetchSubmissionResponse = mockSuccess((submissionUUID) => ({
  response: fakeData.mockSubmission(submissionUUID).response,
}));

/* I assume this is the "Start Grading" call, even for if a
 * submission is already graded and we are attempting re-lock.
 * Assuming the check for if allowed would happen locally first.
 * post('api/lock', { ora_location, submissionUUID });
 * @param {bool} value - new lock value
 * @param {string} submissionUUID
 */
const lockSubmission = mockSuccess(({ submissionUUID, value }) => ({
  ...fakeData.mockSubmissionStatus(submissionUUID),
  lockStatus: value ? lockStatuses.inProgress : lockStatuses.unlocked,
}));

/*
 * Assuming we do not care who has locked it or why, as there
 * is no design around communicating that info
 * post('api/lock', { submissionUUID });
 * @param {bool} value - new lock value
 * @param {string} submissionUUID
 */
const lockSubmissionFail = mockFailure(() => ({
  error: 'that did not work',
}));

/*
 * post('api/updateGrade', { submissionUUID, gradeData })
 * @param {object} gradeData - full grading submission data
 */
const updateGrade = mockSuccess((submissionUUID, gradeData) => ({
  gradeData,
  gradeStatus: gradeStatuses.graded,
  lockStatus: lockStatuses.unlocked,
}));

export default StrictDict({
  initializeApp,
  fetchSubmission,
  fetchSubmissionResponse,
  fetchSubmissionStatus,
  lockSubmission,
  lockSubmissionFail,
  updateGrade,
});
