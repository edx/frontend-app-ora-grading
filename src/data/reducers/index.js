import { combineReducers } from 'redux';

import app from './app';
import submissions from './submissions';

/* istanbul ignore next */
const rootReducer = combineReducers({
  app,
  submissions,
});

export default rootReducer;
