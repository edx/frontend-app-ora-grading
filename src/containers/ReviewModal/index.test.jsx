import React from 'react';
import { shallow } from 'enzyme';

import { formatMessage } from 'testUtils';
import { selectors, actions, thunkActions } from 'data/redux';
import { RequestKeys } from 'data/constants/requests';
import { gradingStatuses as statuses } from 'data/services/lms/constants';

import {
  ReviewModal,
  mapStateToProps,
  mapDispatchToProps,
} from '.';

jest.useFakeTimers('modern');

jest.mock('data/redux', () => ({
  selectors: {
    app: {
      ora: { name: (...args) => ({ oraName: args }) },
      showReview: (...args) => ({ showReview: args }),
    },
    grading: {
      selected: {
        response: (...args) => ({ selectedResponse: args }),
        gradingStatus: (...args) => ({ selectedGradeStatus: args }),
      },
    },
    requests: {
      isCompleted: (...args) => ({ isCompleted: args }),
      errorStatus: (...args) => ({ errorStatus: args }),
    },
  },
  actions: {
    app: {
      setShowReview: jest.fn(),
    },
  },
  thunkActions: {
    app: {
      initialize: jest.fn(),
    },
    grading: {
      cancelGrading: jest.fn(),
    },
  },
}));

jest.mock('containers/ReviewActions', () => 'ReviewActions');
jest.mock('./ReviewContent', () => 'ReviewContent');
jest.mock('components/LoadingMessage', () => 'LoadingMessage');
jest.mock('./components/CloseReviewConfirmModal', () => 'CloseReviewConfirmModal');

const requestKey = RequestKeys.fetchSubmission;

describe('ReviewModal component', () => {
  let el;
  const props = {
    oraName: 'test-ora-name',
    isOpen: false,
    response: { text: (<div>some text</div>) },
    showRubric: false,
    isLoaded: false,
    errorStatus: null,
    gradingStatus: statuses.ungraded,
    intl: { formatMessage },
  };
  beforeEach(() => {
    props.setShowReview = jest.fn();
    props.stopGrading = jest.fn();
    props.reloadSubmissions = jest.fn();
  });
  describe('component', () => {
    describe('snapshots', () => {
      let render;
      beforeEach(() => {
        el = shallow(<ReviewModal {...props} />);
        el.instance().onClose = jest.fn().mockName('this.onClose');
        el.instance().closeModal = jest.fn().mockName('this.closeModal');
        el.instance().showConfirmCloseReviewGrade = jest.fn().mockName('this.showConfirmCloseReviewGrade');
        el.instance().hideConfirmCloseReviewGrade = jest.fn().mockName('this.hideConfirmCloseReviewGrade');
        el.instance().confirmCloseReviewGrade = jest.fn().mockName('this.confirmCloseReviewGrade');
        render = () => el.instance().render();
      });
      test('closed', () => {
        expect(render()).toMatchSnapshot();
      });
      test('loading', () => {
        el.setProps({ isOpen: true });
        expect(render()).toMatchSnapshot();
      });
      test('error', () => {
        el.setProps({ isOpen: true, errorStatus: 200 });
        expect(render()).toMatchSnapshot();
      });
      test('success', () => {
        el.setProps({ isOpen: true, isLoaded: true });
        expect(render()).toMatchSnapshot();
      });
      test('success, demo (title message)', () => {
        const oldEnv = process.env;
        process.env.REACT_APP_NOT_ENABLED = true;
        el.setProps({ isOpen: true, isLoaded: true });
        expect(render()).toMatchSnapshot();
        process.env = oldEnv;
      });
    });

    describe('component', () => {
      describe('close modal', () => {
        test('is not grading', () => {
          el = shallow(<ReviewModal {...props} />);
          el.instance().onClose();
          expect(el.state('showConfirmCloseReviewGrade')).toBe(false);

          const { setShowReview, reloadSubmissions, stopGrading } = props;
          expect(stopGrading).not.toHaveBeenCalled();
          expect(setShowReview).toHaveBeenCalled();
          expect(reloadSubmissions).toHaveBeenCalled();
        });

        describe('is grading', () => {
          beforeEach(() => {
            el = shallow(<ReviewModal {...props} gradingStatus={statuses.inProgress} />);
          });

          test('show modal', () => {
            el.instance().onClose();
            expect(el.state('showConfirmCloseReviewGrade')).toBe(true);
          });

          test('cancel closing then just close confirm do nothing else', () => {
            el.instance().onClose();
            el.instance().hideConfirmCloseReviewGrade();
            expect(el.state('showConfirmCloseReviewGrade')).toBe(false);

            const { setShowReview, reloadSubmissions, stopGrading } = props;
            expect(stopGrading).not.toHaveBeenCalled();
            expect(setShowReview).not.toHaveBeenCalled();
            expect(reloadSubmissions).not.toHaveBeenCalled();
          });

          test('confirm closing then stop grading and close the modal', () => {
            el.instance().onClose();
            el.instance().confirmCloseReviewGrade();
            expect(el.state('showConfirmCloseReviewGrade')).toBe(false);

            const { setShowReview, reloadSubmissions, stopGrading } = props;
            expect(stopGrading).toHaveBeenCalled();
            expect(setShowReview).toHaveBeenCalled();
            expect(reloadSubmissions).toHaveBeenCalled();
          });
        });
      });
    });
  });
  describe('mapStateToProps', () => {
    let mapped;
    const testState = { some: 'test-state' };
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('oraName loads from app.ora.name', () => {
      expect(mapped.oraName).toEqual(selectors.app.ora.name(testState));
    });
    test('isOpen loads from app.showReview', () => {
      expect(mapped.isOpen).toEqual(selectors.app.showReview(testState));
    });
    test('response loads from grading.selected.response', () => {
      expect(mapped.response).toEqual(selectors.grading.selected.response(testState));
    });
    test('isLoaded loads from requests.isCompleted(fetchSubmission)', () => {
      expect(mapped.isLoaded).toEqual(selectors.requests.isCompleted(testState, { requestKey }));
    });
    test('errorStatus loads from requests.errorStatus(fetchSubmission)', () => {
      expect(mapped.errorStatus).toEqual(selectors.requests.errorStatus(testState, { requestKey }));
    });

    test('gradingStatus loads from grading.selected.gradingStatus', () => {
      expect(mapped.gradingStatus).toEqual(selectors.grading.selected.gradingStatus(testState));
    });
  });
  describe('mapDispatchToProps', () => {
    it('loads setShowReview from actions.app.setShowReview', () => {
      expect(mapDispatchToProps.setShowReview).toEqual(actions.app.setShowReview);
    });

    it('loads reloadSubmissions from thunkActions.app.initialize', () => {
      expect(mapDispatchToProps.reloadSubmissions).toEqual(thunkActions.app.initialize);
    });

    it('loads stopGrading from thunkActions.grading.cancelGrading', () => {
      expect(mapDispatchToProps.stopGrading).toEqual(thunkActions.grading.cancelGrading);
    });
  });
});
