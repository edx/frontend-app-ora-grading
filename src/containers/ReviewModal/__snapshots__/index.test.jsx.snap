// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewModal component component snapshots closed 1`] = `undefined`;

exports[`ReviewModal component component snapshots loading 1`] = `undefined`;

exports[`ReviewModal component component snapshots success 1`] = `undefined`;
